cluster-autoscaler/cloudprovider/openstack/openstack_manager_heat.go

```diff
@@ -0,0 +1,299 @@
+	clusterName   string
+
+	stackName string
+	stackID   string
+
+	waitTimeStep time.Duration
+}
+
+// CreateOpenstackManagerHeat sets up cluster and stack clients and returns
+// an OpenstackManagerHeat.
+func CreateOpenstackManagerHeat(configReader io.Reader, discoverOpts cloudprovider.NodeGroupDiscoveryOptions, opts config.AutoscalingOptions) (*OpenstackManagerHeat, error) {
+	var cfg provider_os.Config
+	if configReader != nil {
+		if err := gcfg.ReadInto(&cfg, configReader); err != nil {
+			klog.Errorf("Couldn't read config: %v", err)
+			return nil, err
+		}
+	}
+
+	if opts.ClusterName == "" {
+		klog.Fatalf("The cluster-name parameter must be set")
+	}
+
+	authOpts := toAuthOptsExt(cfg)
+
+	provider, err := openstack.NewClient(cfg.Global.AuthURL)
+	if err != nil {
+		return nil, fmt.Errorf("could not authenticate client: %v", err)
+	}
+
+	if cfg.Global.CAFile != "" {
+		roots, err := certutil.NewPool(cfg.Global.CAFile)
+		if err != nil {
+			return nil, err
+		}
+		config := &tls.Config{}
+		config.RootCAs = roots
+		provider.HTTPClient.Transport = netutil.SetOldTransportDefaults(&http.Transport{TLSClientConfig: config})
+
+	}
+
+	err = openstack.AuthenticateV3(provider, authOpts, gophercloud.EndpointOpts{})
+	if err != nil {
+		return nil, fmt.Errorf("could not authenticate: %v", err)
+	}
+
+	clusterClient, err := openstack.NewContainerInfraV1(provider, gophercloud.EndpointOpts{Type: "container-infra", Name: "magnum", Region: cfg.Global.Region})
+	if err != nil {
+		return nil, fmt.Errorf("could not create container-infra client: %v", err)
+	}
+
+	heatClient, err := openstack.NewOrchestrationV1(provider, gophercloud.EndpointOpts{Type: "orchestration", Name: "heat", Region: cfg.Global.Region})
+	if err != nil {
+		return nil, fmt.Errorf("could not create orchestration client: %v", err)
+	}
+
+	manager := OpenstackManagerHeat{
+		clusterClient: clusterClient,
+		clusterName:   opts.ClusterName,
+		heatClient:    heatClient,
+		waitTimeStep:  waitForStatusTimeStep,
+	}
+
+	manager.stackID, err = manager.GetStackID()
+	if err != nil {
+		return nil, fmt.Errorf("could not store stack ID on manager: %v", err)
+	}
+	manager.stackName, err = manager.GetStackName()
+	if err != nil {
+		return nil, fmt.Errorf("could not store stack name on manager: %v", err)
+	}
+
+	return &manager, nil
+}
+
+// NodeGroupSize gets the current cluster size as reported by magnum.
+// The nodegroup argument is ignored as this implementation of OpenstackManager
+// assumes that only a single node group exists.
+func (osm *OpenstackManagerHeat) NodeGroupSize(nodegroup string) (int, error) {
+	cluster, err := clusters.Get(osm.clusterClient, osm.clusterName).Extract()
+	if err != nil {
+		return 0, fmt.Errorf("could not get cluster: %v", err)
+	}
+	return cluster.NodeCount, nil
+}
+
+// UpdateNodeCount replaces the cluster node_count in magnum.
+func (osm *OpenstackManagerHeat) UpdateNodeCount(nodegroup string, nodes int) error {
```

Comment:
General comment/question about the functions in some of these files:  Could these functions be made private?  I don't think that the core CA code would need to make calls outside of the cloudprovider and nodegroup interfaces (and some of the constructors).

Reply:
Yes, they probably should be.

Reply:
Done in https://github.com/kubernetes/autoscaler/pull/1690/commits/02149e0517bd219e866fccb8bbde766ae16a0104.